### PR TITLE
Extract index sets from particle handlers

### PR DESCRIPTION
**Pull request by luca-heltai**
Extract an IndexSet with global dimensions equal to
<s>`n_comps*particles.get_next_free_particle_index()`</s>, `get_next_free_particle_index()` containing <s>for each
particle index, a set of `n_comps*` consecutive indices associated to
the particles that are locally owned.</s> the locally owned indices.

<s>
The indices associated to a particle with global index `n` will be the
half open range `[n_comps*n`, n_comps*(n+1))`.</s>

### CODE REVIEW

File: source/particles/utilities.cc [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  {
+    template <int dim, int spacedim>
+    IndexSet
+    locally_relevant_ids(const ParticleHandler<dim, spacedim> &particles,
+                         const unsigned int                    n_comps)
+    {
+      IndexSet set(particles.get_next_free_particle_index() * n_comps);
+      for (const auto &p : particles)
+        set.add_range(p.get_id() * n_comps, p.get_id() * n_comps + n_comps);
+      return set;
```
Comment (bangerth) [inline source/particles/utilities.cc]: You'll want a `set.compress()` before the `return`.

**Comment by luca-heltai** [comment]
> It takes n_comp as an input argument, rather than query it from the particles.

The main reason is related to the fact that in principle one may not need properties at all, i.e., if I'm generating a matrix for the interpolation of the particles, the "property" is the position itself (spacedim components) and this has nothing to do with the number of properties.

> My preference would be if you wrote the function so that it really just returns an index set of the locally relevant ids (which is, incidentally, what the function is already called). And then we implement the tensor-product-then-flatten function in IndexSet.

This makes sense. I'll implement it this way then.

> I will note that one way or the other, whatever the function returns, this is going to be a really expensive data structure. IndexSet is optimized to store a relatively small number of ranges, but in parallel computations after some time, particle ids on each processor will become pretty well mixed and not consecutive any more. You'll end up with a lot of small ranges.

Yes indeed. It is as compressed as it can get (and that's why one we want to have a compress particle indices function in the particle handler).

**Comment by bangerth** [comment]
The test currently fails as follows:
```
3594/4062 Test #3598: particles/extract_index_set_01.debug .................................***Failed   12.86 sec

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc: In function ‘void test()’:

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:53:9: error: ‘IteratorFilters’ has not been declared

     tr, IteratorFilters::LocallyOwnedCell());

         ^

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc: In instantiation of ‘void test() [with int dim = 2; int spacedim = 2]’:

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:81:14:   required from here

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:36:72: error: no matching function for call to ‘dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(int)’

   parallel::distributed::Triangulation<dim, spacedim> tr(MPI_COMM_WORLD);

                                                                        ^

In file included from /jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:21:0:

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1416:7: note: candidate: dealii::parallel::distributed::Triangulation<dim, spacedim>::Triangulation() [with int dim = 2; int spacedim = 2] <deleted>

       Triangulation() = delete;

       ^

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1416:7: note:   candidate expects 0 arguments, 1 provided

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note: candidate: dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(const dealii::parallel::distributed::Triangulation<2, 2>&) <deleted>

     class Triangulation

           ^

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note:   no known conversion for argument 1 from ‘int’ to ‘const dealii::parallel::distributed::Triangulation<2, 2>&’

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note: candidate: dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(dealii::parallel::distributed::Triangulation<2, 2>&&) <deleted>

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note:   no known conversion for argument 1 from ‘int’ to ‘dealii::parallel::distributed::Triangulation<2, 2>&&’

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc: In instantiation of ‘void test() [with int dim = 3; int spacedim = 3]’:

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:84:14:   required from here

/jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:36:72: error: no matching function for call to ‘dealii::parallel::distributed::Triangulation<3, 3>::Triangulation(int)’

   parallel::distributed::Triangulation<dim, spacedim> tr(MPI_COMM_WORLD);

                                                                        ^

In file included from /jenkins/workspace/dealii_PR-9074/tests/particles/extract_index_set_01.cc:21:0:

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1416:7: note: candidate: dealii::parallel::distributed::Triangulation<dim, spacedim>::Triangulation() [with int dim = 3; int spacedim = 3] <deleted>

       Triangulation() = delete;

       ^

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1416:7: note:   candidate expects 0 arguments, 1 provided

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note: candidate: dealii::parallel::distributed::Triangulation<3, 3>::Triangulation(const dealii::parallel::distributed::Triangulation<3, 3>&) <deleted>

     class Triangulation

           ^

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note:   no known conversion for argument 1 from ‘int’ to ‘const dealii::parallel::distributed::Triangulation<3, 3>&’

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note: candidate: dealii::parallel::distributed::Triangulation<3, 3>::Triangulation(dealii::parallel::distributed::Triangulation<3, 3>&&) <deleted>

/jenkins/workspace/dealii_PR-9074/include/deal.II/distributed/tria.h:1408:11: note:   no known conversion for argument 1 from ‘int’ to ‘dealii::parallel::distributed::Triangulation<3, 3>&&’

make[3]: *** [CMakeFiles/extract_index_set_01.debug.dir/extract_index_set_01.cc.o] Error 1

make[2]: *** [CMakeFiles/extract_index_set_01.debug.dir/all] Error 2

make[1]: *** [CMakeFiles/extract_index_set_01.debug.diff.dir/rule] Error 2

make: *** [extract_index_set_01.debug.diff] Error 2
```

File: tests/particles/extract_index_set_01.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+#include <deal.II/particles/particle_handler.h>
+
+#include "../tests.h"
+
+template <int dim, int spacedim>
+void
+test()
+{
+  parallel::distributed::Triangulation<dim, spacedim> tr(MPI_COMM_WORLD);
```
Comment (masterleinad) [inline tests/particles/extract_index_set_01.cc]: This requires renaming "tests/particles/extract_index_set_01.output" to "tests/particles/extract_index_set_01.with_p4est=true.mpirun=1.output".

**Pull request merged**